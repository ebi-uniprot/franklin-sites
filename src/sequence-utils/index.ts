export const commentLineRE = /^\s*[>;](.*)$/m;
const whitespacesRE = /\s+/g;

export function formatFASTA(fasta: string, chunkSize = 10, chunksPerLine = 6) {
  const aaPerChunkRE = new RegExp(`(.{1,${chunkSize}})`, 'g');
  const chunksPerLineRE = new RegExp(
    `(.{1,${chunkSize * chunksPerLine}})`,
    'g'
  );

  let header = '';
  let sequence = '';
  for (const line of fasta.split('\n')) {
    const trimmedLine = line.trim();
    if (commentLineRE.test(trimmedLine)) {
      header += `\n${trimmedLine}`;
    } else {
      // concatenate all the sequence lines into a single one
      sequence += trimmedLine;
    }
  }

  let output = sequence
    .replace(whitespacesRE, '')
    // split the sequence per requested aa per lines
    .replace(chunksPerLineRE, '$1\n')
    .split('\n')
    // separate aa per chunk size within the lines themselves
    .map((line) => line.replace(aaPerChunkRE, '$1 ').trim())
    .join('\n');
  if (header) {
    output = `${header}\n${output}`;
  }
  return output.trim();
}

// extract a name as an NCBI identifier from a FASTA header
// See: https://en.wikipedia.org/wiki/FASTA_format#NCBI_identifiers
export function extractNameFromFASTAHeader(fasta?: string) {
  if (!fasta) {
    return;
  }

  const [, header = ''] = fasta.match(commentLineRE) || [];

  // separate by space, first word will be the extracted name
  const [name] = header.split(' ').filter(Boolean);

  // eslint-disable-next-line consistent-return
  return name;
}
