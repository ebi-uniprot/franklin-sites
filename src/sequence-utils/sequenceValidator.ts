const naturalAminoAcids = 'ARNDCEQGHILKMFPSTWYV';
const ambiguousAminoAcids = 'XBZJ';
const baseNucleicAcids = 'ACGTU';
const ambiguousNucleicAcids = 'WSMKRYBDHVNZ';

// Characters unique only to their subset e.g. 'U' is not a valid AA.
const aminoAcidsOnly = /[EQILFPXJ]/gi;
const nucleicAcidsOnly = /U/gi;

export const validAminoAcids = [
  naturalAminoAcids,
  ambiguousAminoAcids,
  '*', // Stop
  '.-', // Gaps
].join('');

export const validNucleicAcids = [
  baseNucleicAcids,
  ambiguousNucleicAcids,
  // No gaps e.g. - and . are not allowed
].join('');

export const errorResponses = Object.freeze({
  missingSequence: Object.freeze({
    valid: false,
    message: 'The sequence is missing',
  }),
  invalidSequence: Object.freeze({
    valid: false,
    message: 'The sequence is invalid',
  }),
  shortSequence: Object.freeze({
    valid: false,
    message: 'The sequence is too short',
  }),
});

export const validResponse = Object.freeze({ valid: true });

// Keep start ^ and end $ anchors in the regex
// Matches all alphabet letters, except for the letter O (case-insensitive)
// Accepts * . -
const validCharacters = /^[A-NP-Z*.-]+$/i;

/**
 * Very basic check on if a string is likely to be a FASTA-formatted string
 *
 * @param {string} seq - Sequence
 * @return {Boolean} True if it is likely to be FASTA
 */
const isFASTA = (seq: string) => /.*[>;]+/gm.test(seq);

/**
 * Prepares a string to be digested by the core validation function.
 *
 * @param {string} fasta - A FASTA-formatted string
 * @return {Array<string>} An array of clean sequences
 */
function prepareFASTAString(fasta: string) {
  return fasta
    .split(/^[>;].*\n?$/gm) // split and remove the 'Description' line
    .map((s) => s.replace(/\s/g, '')) // remove all of the white-space
    .filter(Boolean); // remove all non-truthy values e.g. null, '', false.
}

/*
 * Accepts a sequence of characters and makes a guess if
 * the sequence is an AA or NA sequence.
 * Note: This function will always return a guess.
 * Note: The argument 'threshold' is the frequency of nucliec-acid
 * bases that has to be repeated in the sequence in order to consider
 * it a NA sequence. This is a percentage value, represented in
 * values from 1 to 100.
 * e.g. with 'threshold' set to 90
 * sample sequence: AAAAAZZZZZ -> AA
 * sample sequence: AAAAAAAAAZ -> NA
 *
 * @param {string} sequence - A sequence
 * @param {Number} threshold - NA threshold from 1 to 100
 * @return {string} The likely type either 'aa' or 'na'
 */
function guessSequenceType(sequence: string, threshold: number) {
  const typeAA = 'aa';
  const typeNA = 'na';

  // Check for nucleic-acid only characters
  if (nucleicAcidsOnly.test(sequence)) {
    // Make sure it does NOT contain any amino-acid only characters
    if (!aminoAcidsOnly.test(sequence)) {
      return typeNA;
    }
  }

  // Check for amino-acid only characters
  if (aminoAcidsOnly.test(sequence)) {
    // Make sure it does NOT contain any nucleic-acid only characters
    if (!nucleicAcidsOnly.test(sequence)) {
      return typeAA;
    }
  }

  // Counting the characters to determine the percentage of
  // total occurances of members of each type
  const counts: Record<string, number> = {};

  for (let index = 0; index < sequence.length; index += 1) {
    const char = sequence.charAt(index);

    if (!counts[char]) {
      counts[char] = 1;
    } else {
      counts[char] += 1;
    }
  }

  // Keeping track of each individual character's percentage in the
  // whole sequence, gives flexibility if further logic needs to be
  // utilised in the future.
  const percentagePerChar = 100 / sequence.length;

  const percentages = Object.keys(counts).reduce(
    (acc, current) => {
      if (naturalAminoAcids.includes(current)) {
        acc.aa += counts[current] * percentagePerChar;
      }

      if (baseNucleicAcids.includes(current)) {
        acc.na += counts[current] * percentagePerChar;
      }

      return acc;
    },
    {
      aa: 0, // Amino-Acids
      na: 0, // Nucleic-Acids
    }
  );

  // Is this above our arbitrary threshold?
  if (percentages.na > threshold) {
    return typeNA;
  }

  // If you have reached here and it is not NA, then our guess
  // would be AA.
  return typeAA;
}

/*
 * Does some ground work before passing the sequence to the
 * 'guess' function.
 *
 * @param {string} sequence - A sequence
 * @return {string} The likely type either 'aa' or 'na'
 */
export function findLikelyType(sequence: string) {
  // 1. Remove all of the non-letter characters, plus N and X
  // 2. If less than 11 usable characters left: unable to guess
  // 3. If more than (by default) 90% ACGTU: Nucleic-Acids
  // 4. Else: Amino-Acids
  // Note: Removing N and X is fine because 'N' is valid for both AA and NA sequences,
  // therefore, removing it should equally reduce the sequence length, regardless of the
  // type of sequence. 'X' can be safely removed since it only exists in AA sequences
  // and not a valid character in NA sequences.
  const cleanUpRegEx = /[^A-Z]|[NX]/gi;
  const cleanSequence = sequence.replace(cleanUpRegEx, '');
  const nucleicAcidBaseThreshold = 90;

  return guessSequenceType(cleanSequence, nucleicAcidBaseThreshold);
}

/**
 * Core internal validation function for a single sequence
 *
 * @param {string} sequence - A sequence
 * @return {object} The result
 */
export function sequenceValidator(sequence: string) {
  // Sequence was not passed at all
  if (!sequence) {
    return errorResponses.missingSequence;
  }

  // Remove all white-spaces and FASTA bits
  let cleanSequence: string | undefined;

  if (isFASTA(sequence)) {
    const cleanSequences = prepareFASTAString(sequence);

    if (cleanSequences.length > 0) {
      [cleanSequence] = cleanSequences;
    }
  } else {
    cleanSequence = sequence.replace(/\s/g, '');
  }

  // Nothing left?
  if (!cleanSequence) {
    return errorResponses.missingSequence;
  }

  // Minimum length from: https://www.ebi.ac.uk/ipd/imgt/hla/blast.html
  if (cleanSequence.length < 11) {
    return errorResponses.shortSequence;
  }

  // Check and fail if there are any invalid characters in the sequence
  const onlyValidChar = validCharacters.test(cleanSequence);

  if (!onlyValidChar) {
    return errorResponses.invalidSequence;
  }

  // Attempt to find what type is more likely
  const likelyType: 'na' | 'aa' = findLikelyType(cleanSequence);

  // Fail if type can't be detected
  if (!likelyType) {
    return errorResponses.invalidSequence;
  }

  return {
    ...validResponse,
    likelyType,
    sequence,
  };
}

/**
 * Main validation function
 *
 * @param {Array<string>|string} sequences - An array of sequences or a string
 * @return {Array<object|null>} The result of validating each sequence
 * while keeping order intact.
 */
function validateSequences(input: string | string[]) {
  let sequences: string[] = [];
  // Is this a string?
  if (typeof input === 'string') {
    // Is this a FASTA format?
    if (isFASTA(input)) {
      // This is important to happen here, otherwise you will get a result
      // similar to [[a], [b]] instead of [a, b]
      sequences = prepareFASTAString(input);
    } else {
      // We only work with arrays, so create an array
      sequences = [input];
    }
  } else {
    sequences = [...input];
  }

  // This works based on the value of 'sequence', so keep it here instead of top
  const invalidInputException = `Sequence Validation function expects an Array<string>|string, but received ${typeof sequences}`;

  // Otherwise, make sure we have an array to work with
  if (!Array.isArray(sequences)) {
    throw new Error(invalidInputException);
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
  } else if (!Object.prototype.toString.call(sequences) === '[object Array]') {
    throw new Error(invalidInputException);
  }

  // If the input is empty, return empty results
  if (sequences.length === 0) {
    return [];
  }

  // Validate each sequence separately, compile and return the results
  return sequences.map((sequence) => sequenceValidator(sequence));
}

export default validateSequences;
