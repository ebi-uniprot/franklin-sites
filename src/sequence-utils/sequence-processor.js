import { commentLineRE, extractNameFromFASTAHeader } from '.';
import { sequenceValidator } from './sequenceValidator';

const whitespaceRE = /\s+/g;

const getNewSequenceObject = () => ({
  raw: '',
  name: '',
  header: '',
  sequence: '',
});

const validate = (sequenceObject) => {
  const validation = sequenceValidator(sequenceObject.sequence);
  return { ...validation, ...sequenceObject };
};

const sequenceProcessor = (rawText) => {
  const sequences = [];
  let currentSequence = getNewSequenceObject();
  for (const line of rawText.split('\n')) {
    // for each line

    if (commentLineRE.test(line)) {
      // if this is a comment line

      if (currentSequence.sequence) {
        // if we already have sequence data being processed
        // store current sequence
        sequences.push(validate(currentSequence));

        // and start new sequence
        currentSequence = getNewSequenceObject();
      }

      if (currentSequence.header) {
        // multiline header
        currentSequence.header += '\n';
      } else {
        // first header line
        const name = extractNameFromFASTAHeader(line);
        if (name) {
          currentSequence.name = name;
        }
      }
      currentSequence.header += line;
    } else {
      // if this is a sequence line
      currentSequence.sequence += line.replace(whitespaceRE, '');
    }

    // store the raw string
    if (currentSequence.raw) {
      currentSequence.raw += '\n';
    }
    currentSequence.raw += line;
  }

  if (currentSequence.raw) {
    sequences.push(validate(currentSequence));
  }

  return sequences;
};

export default sequenceProcessor;
