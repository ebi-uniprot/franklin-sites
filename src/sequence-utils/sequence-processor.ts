import { commentLineRE, extractNameFromFASTAHeader } from '.';
import { sequenceValidator } from './sequenceValidator';

const whitespaceRE = /\s+/g;

export type SequenceObject = {
  raw: string;
  name: string;
  header: string;
  sequence: string;
  likelyType?: 'aa' | 'na';
  valid?: boolean;
  message?: string;
};

const getNewSequenceObject = (): SequenceObject => ({
  raw: '',
  name: '',
  header: '',
  sequence: '',
});

const validate = ({
  sequenceObject,
  minimumLength,
  strict,
}: {
  sequenceObject: SequenceObject;
  minimumLength?: number;
  strict?: boolean;
}): SequenceObject => {
  const validation = sequenceValidator(
    sequenceObject.sequence,
    minimumLength,
    strict
  );
  return { ...validation, ...sequenceObject };
};

const sequenceProcessor = (
  rawText: string,
  minimumLength?: number,
  strict?: boolean // Strict validates only conventional amino acids (no ambiguous amino acids)
): SequenceObject[] => {
  const sequences: SequenceObject[] = [];
  let currentSequence = getNewSequenceObject();
  for (const line of rawText.split('\n')) {
    // for each line

    if (commentLineRE.test(line)) {
      // if this is a comment line

      if (currentSequence.sequence) {
        // if we already have sequence data being processed
        // store current sequence
        sequences.push(
          validate({ sequenceObject: currentSequence, minimumLength, strict })
        );

        // and start new sequence
        currentSequence = getNewSequenceObject();
      }

      if (currentSequence.header) {
        // multiline header
        currentSequence.header += '\n';
      } else {
        // first header line
        const name = extractNameFromFASTAHeader(line);
        if (name) {
          currentSequence.name = name;
        }
      }
      currentSequence.header += line;
    } else {
      // if this is a sequence line
      currentSequence.sequence += line.replace(whitespaceRE, '');
    }

    // store the raw string
    if (currentSequence.raw) {
      currentSequence.raw += '\n';
    }
    currentSequence.raw += line;
  }

  if (currentSequence.raw) {
    sequences.push(validate({ sequenceObject: currentSequence, strict }));
  }

  return sequences;
};

export default sequenceProcessor;
