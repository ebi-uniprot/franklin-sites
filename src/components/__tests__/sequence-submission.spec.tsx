import { screen, render, fireEvent } from '@testing-library/react';

import SequenceSubmission from '../sequence-submission';
import { validResponse } from '../../sequence-utils/sequenceValidator';

describe('SequenceSubmission', () => {
  it('should render', () => {
    const { asFragment } = render(<SequenceSubmission />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with sequence is invalid error', () => {
    const { asFragment } = render(
      <SequenceSubmission value="ACTGUACTGUACTGU+" />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with sequence is missing error', () => {
    const { asFragment } = render(<SequenceSubmission value="            " />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should validate after onChange event', async () => {
    render(<SequenceSubmission />);
    const textarea = screen.getByTestId('sequence-submission-input');
    fireEvent.change(textarea, { target: { value: 'ACTGUACTGUACTGU+' } });
    const error = await screen.findByTestId('sequence-submission-error');
    expect(error).toBeInTheDocument();
  });

  it('should call custom onChange callback once', async () => {
    const onChange = jest.fn();
    const value = 'ACTGUACTGUACTGU';
    render(<SequenceSubmission onChange={(e) => onChange(e)} />);
    const textarea = screen.getByTestId('sequence-submission-input');
    fireEvent.change(textarea, { target: { value } });
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith([
      {
        ...validResponse,
        header: '',
        name: '',
        sequence: value,
        likelyType: 'na',
        raw: value,
      },
    ]);
  });
});
