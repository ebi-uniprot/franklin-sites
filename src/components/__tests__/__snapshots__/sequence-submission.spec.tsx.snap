// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SequenceSubmission should render 1`] = `
<DocumentFragment>
  <textarea
    class="sequence-submission-input"
    data-testid="sequence-submission-input"
    spellcheck="false"
  />
</DocumentFragment>
`;

exports[`SequenceSubmission should render with sequence is invalid error 1`] = `
<DocumentFragment>
  <textarea
    class="sequence-submission-input"
    data-testid="sequence-submission-input"
    spellcheck="false"
  >
    ACTGUACTGUACTGU+
  </textarea>
  <div
    class="message message--info"
    role="status"
  >
    <div
      class="message__side-border"
    />
    <test-file-stub
      height="1.125em"
      width="1.125em"
    />
    <section
      class="message__content"
    >
      <small>
        Your input contains 1 sequence
      </small>
    </section>
  </div>
  <div
    class="message message--failure"
    data-testid="sequence-submission-error"
    role="status"
  >
    <div
      class="message__side-border"
    />
    <test-file-stub
      height="1.125em"
      width="1.125em"
    />
    <section
      class="message__content"
    >
      <small>
        <code>
          sequence 1
        </code>
        :Â The sequence is invalid
      </small>
    </section>
  </div>
</DocumentFragment>
`;

exports[`SequenceSubmission should render with sequence is missing error 1`] = `
<DocumentFragment>
  <textarea
    class="sequence-submission-input"
    data-testid="sequence-submission-input"
    spellcheck="false"
  >
                
  </textarea>
</DocumentFragment>
`;
